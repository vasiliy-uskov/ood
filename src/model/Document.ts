import {DocumentItem, ReadonlyDocumentItem} from "./DocumentItem";

export class Document {
	public readonly content: ReadonlyArray<DocumentItem>;
	public readonly title: string;
}

export class ReadonlyDocument {
	public readonly content: ReadonlyArray<ReadonlyDocumentItem>;
	public readonly title: string;
}