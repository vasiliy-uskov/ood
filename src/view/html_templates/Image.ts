import {Size} from "../../utils/Size";

export type ImageProps = {
	readonly src: string,
	readonly size: Size
};

export function Image(props: ImageProps) {
	return `<img src="${props.src}" width="${props.size.width}" height="${props.size.height}" alt="" />`;
}