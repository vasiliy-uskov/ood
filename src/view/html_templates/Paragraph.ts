import {escapeHtmlText} from "../../utils/utils";

export type ParagraphProps = {
	readonly text: string
};

export function Paragraph(props: ParagraphProps) {
	return `<p>${escapeHtmlText(props.text)}</p>`
}