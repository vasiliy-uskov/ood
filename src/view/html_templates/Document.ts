import {escapeHtmlText} from "../../utils/utils";
import {Image, ImageProps} from "./Image";
import {Paragraph, ParagraphProps} from "./Paragraph";

export enum ContentType {
	PARAGRAPH = 'paragraph',
	IMAGE = 'image',
}

export type ParagraphDocumentItem = {
	type: ContentType.PARAGRAPH,
	props: ParagraphProps,
}

export type ImageDocumentItem = {
	type: ContentType.IMAGE,
	props: ImageProps,
}

export type DocumentContent = ReadonlyArray<(ParagraphDocumentItem|ImageDocumentItem)>;

export type DocumentProps = {
	title: string,
	content: DocumentContent,
}

export function Document(props: DocumentProps) {
	return `<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>${escapeHtmlText(props.title)}</title>
</head>
<body>
	${props.content.map(item => {
		switch (item.type) {
			case ContentType.IMAGE:
				return Image(item.props);
			case ContentType.PARAGRAPH:
				return Paragraph(item.props);
			default:
				throw new Error("Undefined content type")
		}
	}).join("")}
</body>
</html>`;
}