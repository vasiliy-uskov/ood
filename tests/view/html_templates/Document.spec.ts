import {ContentType, Document, DocumentContent} from "../../../src/view/html_templates/Document";

function expandDocument(title: string, content: DocumentContent) {
	expect(Document({title, content})).toMatchSnapshot();
}

const jsInjectionText = `<script>alert('&nbsp;')</script>`;

it('expand empty paragraph', () => {
	expandDocument('Empty document', []);
});

it('escape js injection', () => {
	expandDocument(jsInjectionText, []);
});

it('expand not empty paragraph', () => {
	expandDocument('Not empty', [
		{
			type: ContentType.PARAGRAPH,
			props: {
				text: jsInjectionText,
			}
		},
		{
			type: ContentType.IMAGE,
			props: {
				src: './location/src.png',
				size: {
					width: 100,
					height: 100,
				}
			}
		}
	]);
});

it('throw error on wrong content type', () => {
	expect(() => Document({
		title: '',
		content: [{
			type: 'Some wrong content type',
		}] as any as DocumentContent
	})).toThrow();
});