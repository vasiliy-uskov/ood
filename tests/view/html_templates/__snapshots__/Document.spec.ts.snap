// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`escape js injection 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
	<meta charset=\\"UTF-8\\">
	<title>&lt;script&gt;alert('&amp;nbsp;')&lt;/script&gt;</title>
</head>
<body>
	
</body>
</html>"
`;

exports[`expand empty paragraph 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
	<meta charset=\\"UTF-8\\">
	<title>Empty document</title>
</head>
<body>
	
</body>
</html>"
`;

exports[`expand not empty paragraph 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
	<meta charset=\\"UTF-8\\">
	<title>Not empty</title>
</head>
<body>
	<p>&lt;script&gt;alert('&amp;nbsp;')&lt;/script&gt;</p><img src=\\"./location/src.png\\" width=\\"100\\" height=\\"100\\" alt=\\"\\" />
</body>
</html>"
`;
